_posts/2018-08-04-pfsense-as-hvm-on-pc-engines-apu2.md

```diff
@@ -0,0 +1,574 @@
+installation.
+
+## PCI passthrough in Debian
+
+Below screenshot show device `02:00.0`, which is apu2 middle NIC,
+passed-through to VM.
+
+![Debian lspci](https://3mdeb.com/wp-content/uploads/2017/07/debian-9.5.0-hvm-pci-passthrough.png)
+
+PCI passthrough on Debian worked without any issue DHCP offer was received
+correctly and I could proceed with performance checks.
+
+# Speedtest
+
+Simplest possible test is comparison of through put between eth0 and eth1.
+First is connected directly to our company switch and second connects pfSense
+HVM using PCI passthrough.
+
+I used `speedtest-cli v2.0.2`.
+
+Results for apu2 Dom0:
+
+```
+(speedtest-venv) root@apu2:~# speedtest-cli 
+Retrieving speedtest.net configuration...
+Testing from Vectra Broadband (109.241.231.46)...
+Retrieving speedtest.net server list...
+Selecting best server based on ping...
+Hosted by Volta Communications Sp. z o.o (Gdansk) [2.28 km]: 36.105 ms
+Testing download speed................................................................................
+Download: 81.67 Mbit/s
+Testing upload speed................................................................................................
+Upload: 15.38 Mbit/s
+```
+
+Results for Debian HVM with NIC PCI passthrough:
+
+![Debian HVM speedtest-cli](https://3mdeb.com/wp-content/uploads/2017/07/speedtest-cli-debian-hvm.png)
+
+# iperf
+
+Below results are for very simple LAN connection `apu3 -> switch -> apu2`:
+
+```
+(speedtest-venv) root@apu2:~# iperf -s -B 192.168.3.101
+------------------------------------------------------------
+Server listening on TCP port 5001
+Binding to local address 192.168.3.101
+TCP window size: 85.3 KByte (default)
+------------------------------------------------------------
+[  4] local 192.168.3.101 port 5001 connected with 192.168.3.102 port 34004
+[ ID] Interval       Transfer     Bandwidth
+[  4]  0.0-10.0 sec  1.10 GBytes   941 Mbits/sec
+```
+
+Unfortunately our switch is probably not well suited for testing 1GbE. Those
+tests should be repeated with directly connected ports/devices.
+
+Results for Debian HVM with NIC PCI passthrough:
+
+![Debian HVM iperf](https://3mdeb.com/wp-content/uploads/2017/07/1GbE.png)
+
+As you can see there is no difference between results, based on that we can
+conclude that PCI passthrough works and there is no overhead when using IOMMU.
+
+Below log show results from Debian PV and prove how virtualized drivers lead to
+performance overhead.
+
+```
+root@debian-pv:~# iperf -c 192.168.3.128
+------------------------------------------------------------
+Client connecting to 192.168.3.128, TCP port 5001
+TCP window size: 85.0 KByte (default)
+------------------------------------------------------------
+[  3 ] local 192.168.3.105 port 56204 connected with 192.168.3.128 port 5001
+[ ID ] Interval       Transfer     Bandwidth
+[  3 ]  0.0-10.0 sec   746 MBytes   625 Mbits/sec
+```
+
+# Possible problems
+
+# xen-pciback not loaded
```

Comment:
@pietrushnic I think those subsections describing each problem should have lower header level, i.e. `##`

Reply:
@miczyg1 agree, done